Added functionality to log why a task will be rebuild.

felixfontein opened the pull request:
Addressing #24.

Will show the reason why a task is rebuild on `doit info`.

felixfontein commented:
The last commit is kind of a hack to get the unit tests working again. They failed because the `CmdInfo` object had `dep_manager` not set. This decreases the coverage (because the new code isn't tested in any unit test).

schettino72 commented:
Thanks for working on this...

- what about adding a command line option? IMO this information should be displayed only when explicitly asked for.

- It would be nice to display **all** reasons a task is not up-to-date, right now it just gives the first reason found.

- It would be nice if the exit value of the command indicates if the task is up-to-date or not. zero for up-to-date, 1 for not up-to-date.

schettino72 commented on doit/dependency.py:
```diff
@@ -528,7 +528,7 @@ def status_is_ignore(self, task):
         return self._get(task.name, "ignore:")
 
     # TODO add option to log this
```
it is not a TODO anymore :)

felixfontein replied:
Now that it lists all reasons, definitely. Removed it.

schettino72 commented on doit/dependency.py:
```diff
@@ -528,7 +528,7 @@ def status_is_ignore(self, task):
 
     # TODO add option to log this
-    def get_status(self, task, tasks_dict):
+    def get_status(self, task, tasks_dict, rebuild_log=None):
```
that is so ugly. `rebuild_log` is a return value not a parameter. This is not a public API. I make sure doit has 100% code coverage to give confidence in making this kind of changes.

schettino72 commented on doit/dependency.py:
```diff
@@ -593,17 +597,23 @@ def get_status(self, task, tasks_dict):
             if not os.path.exists(targ):
                 task.dep_changed = list(task.file_dep)
+                if rebuild_log is not None:
+                    rebuild_log.append('Target file "{}" does not exist.'.format(targ))
```
It would be better to not repeat the text of the message all the time. I mean I prefer a message like

Target file does not exist:
 - target1
 - target2
 - ...

schettino72 commented on doit/dependency.py:
```diff
@@ -625,11 +640,16 @@ def get_status(self, task, tasks_dict):
             except OSError:
+                if rebuild_log is not None:
+                    rebuild_log.append('Dependent file "{}" does not exist.'.format(dep))
                 raise DependencyException("Dependent file '{}' does not exist."
```
Since rebuild_log will be returned we need to find another solution for this case instead of raising an exception.

schettino72 commented:
Thinking over I think `get_status()` should return a dict (or namedtuple, or custom object) with:

 - result: a string one of `run`, `up-to-date`, `error` (like missing file_dep)
 - missing_file_dep: a list of file_dep names
 - changed_file_dep: list
 - missing_target: list
 - ...

what do you think?

felixfontein commented:
I now added a result object and moved the result formatting to `cmd_info`. Also all results are displayed, and all that is only done when explicitly requested via `-R` or `--show-build-reason`.

About the exit value: what should the exit value be in case `-R` et al is not specified? Also 0?

tom-varga commented:
Will this also work along with --dry-run?
I hope so.

On Wed, Jul 15, 2015 at 3:56 PM, Felix Fontein <notifications@github.com>
wrote:

> I now added a result object and moved the result formatting to cmd_info.
> Also all results are displayed, and all that is only done when explicitly
> requested via -R or --show-build-reason.
>
> About the exit value: what should the exit value be in case -R et al is
> not specified? Also 0?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pydoit/doit/pull/75#issuecomment-121727309>.
>

felixfontein commented:
BTW (completely off-topic), when do you plan to release doit 0.29.0?

felixfontein commented:
@tom-varga: why shouldn't it? It should not affect `doit run` at all.

schettino72 commented on doit/cmd_info.py:
```diff
@@ -23,6 +23,16 @@ def my_safe_repr(obj, context, maxlevels, level):
+opt_show_build_reason = {
+    'name': 'show_build_reason',
+    'short': 'R',
+    'long': 'show-build-reason',
```
IMO this name is too long. Maybe just `status` would be ok. The short name `R` stands for what? and why a capital letter?

felixfontein replied:
I chose `R` for `reason` (I didn't want to use lowecase `r` because I tend to associate that with `recursive`). But `status` sounds good, too, with shortcut `s`.

felixfontein replied:
Changed to `status`/`s`.

schettino72 commented on doit/cmd_info.py:
```diff
@@ -23,6 +23,16 @@ def my_safe_repr(obj, context, maxlevels, level):
+    'long': 'show-build-reason',
+    'type': bool,
+    'default': False,
+    'help': """Shows reasons why this target will be rebuild. [default: %(default)s]"""
```
r/target will be rebuild/task would be executed

felixfontein replied:
Makes sense. I was thinking too much in Nikola terms :)

felixfontein replied:
Changed.

schettino72 commented on doit/cmd_info.py:
```diff
@@ -57,3 +69,41 @@ def _execute(self, pos_args):
+
+        if show_build_reason:
+            if self.dep_manager is None:
+                # Avoid failure in tests which don't initialize dep_manager.
```
Is this still required?

felixfontein replied:
Shouldn't be. Removed.

schettino72 commented on doit/dependency.py:
```diff
@@ -536,13 +555,27 @@ def get_status(self, task, tasks_dict):
         """
+        result = DependencyStatus()
+        if get_log:
+            result.uptodate_false = []
```
Since you created a class `DependencyStatus` this initialization should be done in `DependencyStatus.__init__()`

felixfontein replied:
Done.

schettino72 commented on doit/dependency.py:
```diff
@@ -426,6 +426,26 @@ def get_state(self, dep, current_state):
+    def __init__(self):
+        self.status = None
+
+    def _decide(self, status):
```
Why this method has an `_` ? It is used outside of the class.

felixfontein replied:
Would you prefer `__`?

felixfontein replied:
I now made the `decide()` method local to `get_status` (instead of having it in `DependencyStatus`). Now the only drawback is that `get_status` sets `DependencyStatus._error_message` (which is read by `DependencyStatus.get_error_message()`.

felixfontein replied:
r/`_error_message`/`_error_reason`. And never mind, got rid of underscores here by now.

schettino72 commented on doit/dependency.py:
```diff
@@ -426,6 +426,26 @@ def get_state(self, dep, current_state):
+    def raise_exception(self):
+        """Raises a DependencyException if the state is 'error'."""
+        if self.status == "error":
+            raise DependencyException("Dependent file '{}' does not exist.".format(self._error_reason))
```
I mean this exception could be completely removed. Where this exception is used, it could just deal with the return value.

felixfontein replied:
That's true, but that might be a more complicated transformation, to make sure nothing breaks.

felixfontein replied:
Ok, I tried that now.

schettino72 commented:
> BTW (completely off-topic), when do you plan to release doit 0.29.0?

Normally I would not do a release now since there is so few changes since last release and there are few issues WIP. Anyway if you really need a release there is no problem to push one on your request ;)

felixfontein commented:
It would be nice to have a released version with target regex support and earlytask target specifying on command line. Especially to get the tests in getnikola/nikola#1632 working again. But it's not really that urgent, as I'm not sure when that branch can be merged after all; if there will be another 7.x version of Nikola before Nikola 8 is released, it must wait a bit more.

felixfontein commented:
(Just checked, Nikola 7.6.2 should be released probably this month.)

Oh, and I'll work on the changes in this thread tomorrow, no time left for today :)

felixfontein commented:
Any idea why this one test is failing? I cannot see a connection to my changes.

Kwpolska commented:
Don't expect Nikola v8 to be released soon.

felixfontein commented:
I added some more tests for the `get_log=True` option of `get_status` as well as for `cmd_info`; now the coverage is back to 100%.

schettino72 commented:
thanks. I already started to work on this. so, no need for you to make more commits...

felixfontein commented:
Ok. Thanks!

schettino72 commented:
merged, thanks (rebased it so github cant detect it)

Sorry, I was half-way merging and doing changes when you did your last batch of commits.

I did some changes to try to make get_status simpler because it was already a big function,
so I tried to remove as much complexity as possible from it.

schettino72 commented:
the test failed on travis is #7, hopefully we can get rid of it after #51 is merged.

felixfontein commented:
Ah, ok. I assumed you probably have seen it fail before.
